pkg/kube/configmapwatcher/configmapwatcher.go

```diff
@@ -0,0 +1,137 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configmapwatcher
+
+import (
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	infomersv1 "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"istio.io/istio/pkg/kube"
+	"istio.io/pkg/log"
+)
+
+// Controller watches a ConfigMap and calls the given callback when the ConfigMap changes.
+type Controller struct {
+	client   kube.Client
+	informer infomersv1.ConfigMapInformer
+	queue    workqueue.RateLimitingInterface
```

Comment:
Do we need a queue? Generally we use the queue because we are watching things like all Services in the cluster and doing expensive computations on them. It seems here we are just calling a callback and the queue may be over complicated here?

Reply:
I'm not sure how expensive the installed callbacks are. The queue has logic so that only one item with a given key is processed at the same time, so it can help depending on how expensive they are. I wouldn't mind removing it, although in this case having the queue doesn't add too much code